Support LIST(SUBQUERY) syntax

cjubb39 opened the pull request:
### Motivation
Fixes https://github.com/MaterializeInc/materialize/issues/3878

### Description
- Add support for `LIST(<subquery>)` syntax.  Note that this is different than the `LIST[1,2,3]` syntax used to create a list literal.
- Implement by 1) wrapping each result from the selected column in the subquery into a list and then 2) concatenating those lists.  [This was the approach I came up with because of the particular ways `HirRelationExpr`s can be converted to `HirScalarExpr`s].

### Checklist

- [x] This PR has adequate test coverage.
- [x] This PR adds a release note for any user-facing behavior changes.

sploiselle commented:
From Slack:

> what `RecordCreate` is doing / why I need it

`Record` is our internal type for `Row`-like data. It differs from a `Row` in that it also has the names of the columns it contains, so they're name addressable instead of only on indexes. For example, your call to `RecordCreate` could get the actual column name of the single returned column and use that. Totally unnecessary, but just for a mental model.

So by calling `RecordCreate` on this column, we iterate over the column and turn all of its values into `Records`, so what we have would look like:

```
Record([List([Int32(1)])])
Record([List([Int32(2)])])
```

This data then gets passed to `ListConcat`, which can only work on `Record` data (i.e. instead of taking individual datums directly and packing them into a list) due to something in Differential, whose invariants can be challenging for mere mortals to keep paged in.

philip-stoev commented:
@cjubb39 do mind if I test this? Let me know if you need to push additional commits/fixes into the branch before I start.

cjubb39 commented:
@philip-stoev I think it might make sense to fix the error Sean pointed out before doing some randomized testing.  I'm working on that now!

cjubb39 commented:
@sploiselle I believe I've ironed out the edge cases we've chatted about and I'm happy with this version!  ready for review

@philip-stoev ready for randomized testing!

cjubb39 commented:
I've just pushed changes to address philip's issues:
(1) query plan: not yet addressed
(2) matching postgresql: not in scope here
(3) handling when record create is `eval`d: have a solution in f29a6ed25cea9a50a81dfbffa70a6e2702c4183f but am not convinced it's the best one.
(4) ensuring empty list: fixed with 6331cfe04e26aeeb0e6d52ef8e9a57f28fabb698

wangandi commented:
I consider (1) to be an instance of #8574 and out of the scope of this PR.

For example, this query not involving a list aggregation also has the duplicated NULL value issue:
```
materialize=> explain plan for select (select sum(f1) from t1);
          Optimized Plan           
-----------------------------------
 %0 = Let l0 =                    +
 | Get materialize.public.t1 (u24)+
 | Project (#0)                   +
 | Reduce group=()                +
 | | agg sum(#0)                  +
                                  +
 %1 =                             +
 | Get %0 (l0)                    +
 | Project ()                     +
 | Negate                         +
                                  +
 %2 =                             +
 | Constant ()                    +
                                  +
 %3 =                             +
 | Union %1 %2                    +
 | Map null                       +
                                  +
 %4 = Let l1 =                    +
 | Union %0 %3                    +
                                  +
 %5 =                             +
 | Get %4 (l1)                    +
 | Project ()                     +
 | Negate                         +
                                  +
 %6 =                             +
 | Constant ()                    +
                                  +
 %7 =                             +
 | Union %5 %6                    +
 | Map null                       +
                                  +
 %8 =                             +
 | Union %4 %7                    + 
```

cjubb39 commented:
@sploiselle @aalexandrov @philip-stoev I believe I am all wrapped up and ready for final review!

Addressing Philip's issues:
(1) query plan: not in scope here.  Tracked by issue Andi linked above
(2) matching postgresql: not in scope here
(3) handling when record create is eval'd: fixed with f855864f4426d95991e153b3f3cf7efb9e22fb2a.  This mirrors the solution used in #8939.
(4) ensuring empty list: fixed with b66569024ab1c70d7aa8b8c3403609d8b94e4f61 f855864f4426d95991e153b3f3cf7efb9e22fb2a

sploiselle commented on test/sqllogictest/list_subquery.slt:
```diff
@@ -0,0 +1,162 @@
+----
+{{{alice,lasta},{10003},{1}},{{charlie,lastc},{11217},{3}}}
+
+statement ok
```
minor/unnecessary, but I'm not sure what these tests are stressing––maybe a comment if you end up going back through CI?

sploiselle reviewed (approved):
So cool. Glad you got such a comprehensive tour of SQLland.

cjubb39 commented:
@philip-stoev can you take another look?  I believe I've addressed all the issues you pointed out that are in scope.

cjubb39 commented:
also cc @aalexandrov as a reviewer but I no longer have changes outside the sql parser / planner so I imagine you're probably less interested now.  (The changes I had in the optimizer were addressed in #8973)

aalexandrov commented:
OK, LGTM.

aalexandrov commented on src/sql/src/plan/query.rs:
```diff
@@ -2870,6 +2870,103 @@ pub fn plan_expr<'a>(
         }
+        Expr::ListSubquery(query) => {
+            if !ecx.allow_subqueries {
+                bail!("{} does not allow subqueries", ecx.name)
```
This is either (1) the case also in Postgres or (2) a temporary restriction that needs follow-up work? If (2), can you reference an issue to track this (create one if needed).

cjubb39 replied:
I'm not aware of it being a temporary restriction.  I've included the `ecx.allow_subqueries` check because the other subquery expressions (`Exists` and `Subquery` above) include it.

Do you know why we would not want to check that condition here?

aalexandrov replied:
I guess I was tired and didn't appreciate the meaning of `!ecx.allow_subqueries` guard yesterday. Subqueries are conditionally disallowed depending on the context in which they appear.

aalexandrov commented:
LGTM in general.

However, Sean's detailed description of why we need the `RecordCreate` call does not align with the comment on top of the `ListConcat` variant:

```rust
/// Accumulates `Datum::List`s whose first element is a `Datum::List` into a
/// single `Datum::List`. The other elements are columns used by `order_by`.
ListConcat {
    order_by: Vec<ColumnOrder>,
},
```

I think this needs to be fixed to

```rust
/// Accumulates `Datum::List`s of `ScalarType::Record` whose first field is a `Datum::List` 
/// into a single `Datum::List` and whose remaining fields are used by `order_by`.
ListConcat {
    order_by: Vec<ColumnOrder>,
},
```

It is a bit confusing that the `Datum::List` variant can be typed both as `ScalarType::Record` and `ScalarType::List`.

cjubb39 commented:
chatted in slack in philip.  There have been no major changes since he last evaluated this PR so merging without him re-running randomized testing works
